src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs

```diff
@@ -127,56 +114,75 @@ private void RunBlazorStartupScripts(AWebView view)
-            if (event.data != 'capturePort') {
-                nativeJsPortOne.postMessage(event.data)
-            }
-            else if (event.data == 'capturePort') {
-                if (event.ports[0] != null) {
-                    nativeJsPortTwo = event.ports[0]
-                }
-            }
-        }, false);
-
-        nativeJsPortOne.addEventListener('message', function (event) {
-        }, false);
-
-        nativeJsPortTwo.addEventListener('message', function (event) {
-            // data from native code to JS
-            if (window.external.__callback) {
-                window.external.__callback(event.data);
-            }
-        }, false);
-        nativeJsPortOne.start();
-        nativeJsPortTwo.start();
-
-        window.external.sendMessage = function (message) {
-            // data from JS to native code
-            nativeJsPortTwo.postMessage(message);
-        };
-
-        window.external.receiveMessage = function (callback) {
-            window.external.__callback = callback;
-        }
-
-        ", new JavaScriptValueCallback(() =>
+		const channel = new MessageChannel();
+		var nativeJsPortOne = channel.port1;
+		var nativeJsPortTwo = channel.port2;
+		window.addEventListener('message', function (event) {
+			if (event.data != 'capturePort') {
+				nativeJsPortOne.postMessage(event.data)
+			}
+			else if (event.data == 'capturePort') {
+				if (event.ports[0] != null) {
+					nativeJsPortTwo = event.ports[0]
+				}
+			}
+		}, false);
+
+		nativeJsPortOne.addEventListener('message', function (event) {
+		}, false);
+
+		nativeJsPortTwo.addEventListener('message', function (event) {
+			// data from native code to JS
+			if (window.external.__callback) {
+				window.external.__callback(event.data);
+			}
+		}, false);
+		nativeJsPortOne.start();
+		nativeJsPortTwo.start();
+
+		window.external.sendMessage = function (message) {
+			// data from JS to native code
+			nativeJsPortTwo.postMessage(message);
+		};
+
+		window.external.receiveMessage = function (callback) {
+			window.external.__callback = callback;
+		}
+
+		", new JavaScriptValueCallback(() =>
 			{
 				// Set up Server ports
 				_webViewHandler?.WebviewManager?.SetUpMessageChannel();
 
 				// Start Blazor
 				view.EvaluateJavascript(@"
-                    Blazor.start();
-                ", new JavaScriptValueCallback(() =>
+					Blazor.start();
+				", new JavaScriptValueCallback(() =>
 				{
 					// Done; no more action required
 				}));
 			}));
 		}
 
+		private static bool IsAppOriginPageUri(string? requestUriString)
+		{
+			if (string.IsNullOrEmpty(requestUriString))
+			{
+				return false;
+			}
+
+			var appBaseUri = new Uri(AppOrigin);
+			var requestUri = new Uri(requestUriString);
+
+			if (!appBaseUri.IsBaseOf(requestUri))
+			{
+				return false;
+			}
+
+			// If the path does not end in a file extension, it's most likely referring to a page.
+			return !Path.HasExtension(requestUriString);
```

Comment:
Do we have to get the relative uri first?

```c#
var relativePath = new Uri(AppOrigin).MakeRelativeUri(requestUri).ToString()
```

Also, is the extension check sufficient? What if it's `0.0.0.0/index.html`?

Reply:
@Eilon do you think it would be worth it to special-case the host page path and treat it as a navigation request, or is that overkill?

Reply:
Hmm I don't consider supporting direct navigation to `index.html` important. I kind of see the name `index.html` as an implementation detail because you can't have a file with no name to be at `/`, so we had to name it _something_, but we don't need you to use the name at any other time 😄